core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot.go

```diff
@@ -139,57 +155,261 @@ func newSnapshotWriter(
+	}
+	if metadataFile, err = snapshot.OpenFile(metadataFilePath, snapshotFileFormat); err != nil {
+		return nil, byte(0), errors.WithMessage(err, "error while opening metadata file")
+	}
+	dbValueFormat, err := dataFile.DecodeBytes()
+	if err != nil {
+		return nil, byte(0), errors.WithMessage(err, "error while reading dbvalue-format")
+	}
+	if len(dbValueFormat) != 1 {
+		err = errors.Errorf("dbValueFormat is expected of length  one byte. Found [%d] length", len(dbValueFormat))
+		return nil, byte(0), err
+	}
+
+	metadata, err := loadMetadata(metadataFile)
+	if err != nil {
+		return nil, byte(0), err
+	}
+	return &snapshotReader{
+		dataFile: dataFile,
+		cursor: &cursor{
+			metadata: metadata,
+		},
+	}, dbValueFormat[0], nil
+}
+
+func loadMetadata(metadataFile *snapshot.FileReader) ([]*metadataRow, error) {
+	numMetadata, err := metadataFile.DecodeUVarInt()
+	if err != nil {
+		return nil, errors.WithMessage(err, "error while reading num-rows in metadata")
+	}
+	metadata := []*metadataRow{}
+	for i := uint64(0); i < numMetadata; i++ {
+		ns, err := metadataFile.DecodeString()
+		if err != nil {
+			return nil, errors.WithMessage(err, "error while reading namespace name")
+		}
+		numKVs, err := metadataFile.DecodeUVarInt()
+		if err != nil {
+			return nil, errors.WithMessagef(err, "error while reading num entries for the namespace [%s]", ns)
+		}
+		metadata = append(metadata, &metadataRow{
+			namespace: ns,
+			kvCounts:  numKVs,
+		})
+	}
+	return metadata, nil
+}
+
+func (r *snapshotReader) Next() (*statedb.CompositeKey, []byte, error) {
+	if !r.cursor.move() {
+		return nil, nil, nil
+	}
+
+	key, err := r.dataFile.DecodeString()
+	if err != nil {
+		return nil, nil, errors.WithMessage(err, "error while reading key from datafile")
+	}
+	dbValue, err := r.dataFile.DecodeBytes()
+	if err != nil {
+		return nil, nil, errors.WithMessage(err, "error while reading value from datafile")
+	}
+	return &statedb.CompositeKey{
+		Namespace: r.cursor.currentNamespace(),
+		Key:       key,
+	}, dbValue, nil
+}
+
+func (r *snapshotReader) Close() {
+	if r == nil {
+		return
+	}
+	r.dataFile.Close()
+}
+
+func (r *snapshotReader) hasMore() bool {
+	return r.cursor.canMove()
+}
+
+// metadataRow captures one tuple <namespace, number-of-KVs> in the metadata file
+type metadataRow struct {
+	namespace string
+	kvCounts  uint64
+}
+
+type cursor struct {
+	metadata   []*metadataRow
+	currentRow int
+	movesInRow uint64
+}
+
+func (c *cursor) move() bool {
+	if c.movesInRow < c.metadata[c.currentRow].kvCounts {
+		c.movesInRow++
+		return true
+	}
+	if c.currentRow < len(c.metadata)-1 {
+		c.currentRow++
+		c.movesInRow = 1
+		return true
+	}
+	return false
+}
+
+func (c *cursor) canMove() bool {
+	return c.movesInRow < c.metadata[c.currentRow].kvCounts ||
+		c.currentRow < len(c.metadata)-1
+}
+
+func (c *cursor) currentNamespace() string {
+	return c.metadata[c.currentRow].namespace
+}
```

Comment:
Not sure how it makes it better... I'll prefer to leave this as is.

Reply:
Okay. I respect your decision but I disagree. I only suggested the above for the following reasons:

1. `row`, `moveInRow`, `metadata` etc... are too generic at this level while this cursor is used only for a single purpose. Instead explicit names such as `nsKVCounts`, `indexToCurrentNs`, and `indexToCurrentKV` make it easily relate to the code that uses the cursor. Otherwise, I had to do a mental map while reading the code. 
2. `c.movesInRow < c.metadata[c.currentRow].kvCounts` and `c.currentRow < len(c.metadata)-1` are used twice and hence, it is good to put them in a function called `isCurrentNsNotExhausted()` and `isCursorNotAtLastNs()`. This function name exactly specify what the code does. I know what you would respond for this -- I am using a function name instead of a comment but I do not see anything wrong in that as long as it improves the readability.
3. While this is nit, switch case is suitable here.

Reply:
I would like to agree to disagree on this one